dali_tf_plugin/dali_dataset_op.cc

```diff
@@ -210,7 +270,189 @@ class DALIDatasetOp::Dataset::Iterator : public DatasetIterator<Dataset> {
+      ReleaseInputs();
+    }
+    // We schedule next run always when we don't have inputs or when we have inputs
+    // and they produced something - which happens only in `in_progress` state.
+    if (!dataset()->HasInputs() || iterator_state_ == InputState::in_progress) {
+      TF_DALI_CALL(daliRun(&pipeline_handle_));
+    }
+    return Status::OK();
+  }
+
+  ~Iterator() {
+    if (enable_memory_stats_) {
+      size_t N;
+      daliExecutorMetadata *meta;
+      daliGetExecutorMetadata(&pipeline_handle_, &meta, &N);
+      std::cout << "DALI operator memory statistics: " << std::endl;
+      for (size_t i = 0; i < N; ++i) {
+        std::cout << "Operator " << meta[i].operator_name;
+        for (size_t j = 0; j < meta[i].out_num; ++j) {
+          std::cout << "   output [ " << j << " ] : " << meta[i].real_size[j] << "B allocated "
+                    << meta[i].max_real_size[j] << "B max allocated " << meta[i].reserved[j]
+                    << "B reserved" << meta[i].max_reserved[j] << "B max reserved";
+          if (j != meta[i].out_num - 1) {
+            std::cout << ",";
+          }
+        }
+        std::cout << std::endl;
+      }
+      daliFreeExecutorMetadata(meta, N);
+    }
+    daliDeletePipeline(&pipeline_handle_);
+  }
+
+#if TF_MAJOR_VERSION == 2 && TF_MINOR_VERSION >= 3
+  Status SaveInternal(SerializationContext *ctx, IteratorStateWriter *writer) override {
+    return errors::Unimplemented("SaveInternal is not supported for DALI dataset.");
+  }
+
+  Status RestoreInternal(IteratorContext *ctx, IteratorStateReader *reader) override {
+    return errors::Unimplemented("RestoreInternal is not supported for DALI dataset");
+  }
+#endif
+
+ private:
+  // TODO(klecki): Check this again
+  // TensorFlow treats a sample/example as a vector of Tensors (they flatten everything),
+  // and if a dataset has multiple outputs it means, that it returned a tuple that maps
+  // to a vector of Tensors.
+  using TfExample = std::vector<Tensor>;
+
+  // Batch is a list of Tensors - we repack the TfExamples into a Batch.
+  // We need to build batches sample by sample to advance the input iterators in sync.
+  // TODO(klecki): In batch mode this will be a single Tensor with additional dimension representing
+  //               whole batch
+  using Batch = std::vector<Tensor>;
+
+  // Represents tuple of inputs for one iteration
+  using ListOfBatches = std::vector<Batch>;
+
+
+  /**
+   * @brief Schedule required number of runs of DALI Pipeline to fill the prefetch queue.
+   *
+   * When there are input datasets, feed the pipeline required number of input batches.
+   *
+   * TODO(klecki): Inputs handled only for an uniform executor
+   */
+  Status PrefetchPipeline(IteratorContext *context, daliPipelineHandle *pipeline_handle) {
+    if (!dataset()->pipeline_def_.exec_separated) {
+      int prefetch_depth = dataset()->pipeline_def_.prefetch_queue_depth;
```

Comment:
suggestion: you could define this outside of the if, and reuse later.

Reply:
For what? It's only used in the scope of the if.

Reply:
true, I got confused.